File: packages/server/src/bootstrap/afd-schema.ts
/**
 * @fileoverview afd-schema bootstrap command
 *
 * Export input schemas for all commands as JSON Schema or TypeScript types.
 */

import { success } from '@lushly-dev/afd-core';
import { z } from 'zod';
import { defineCommand, type ZodCommandDefinition } from '../schema.js';
import {
	type DescribedCommand,
	describableCommands,
	type GetDescribedCommands,
} from './described-command.js';
import { generateInputTypes } from './json-schema-to-ts.js';

const inputSchema = z.object({
	format: z
		.enum(['json', 'typescript'])
		.default('json')
		.describe('Output format: JSON Schemas, or TypeScript input types in addition'),
});

interface SchemaInfo {
	name: string;
	description: string;
	inputSchema: Record<string, unknown>;
}

interface SchemaOutput {
	schemas: SchemaInfo[];
	count: number;
	format: 'json' | 'typescript';
	/** With `format: 'typescript'`: one module declaring an input type per command. */
	typescript?: string;
}

/** Build a basic JSON Schema from core `parameters`. */
function schemaFromParameters(cmd: DescribedCommand): Record<string, unknown> {
	const parameters = cmd.parameters ?? [];
	return {
		type: 'object',
		properties: Object.fromEntries(
			parameters.map((p) => [
				p.name,
				{
					type:
						p.type === 'string'
							? 'string'
							: p.type === 'number'
								? 'number'
								: p.type === 'boolean'
									? 'boolean'
									: 'any',
					description: p.description,
				},
			])
		),
		required: parameters.filter((p) => p.required).map((p) => p.name),
	};
}

/**
 * Create the afd-schema bootstrap command.
 *
 * Exports only MCP-exposed commands.
 *
 * @param getCommands - Function to get all registered commands
 * @param getJsonSchema - Function to get JSON schema for a command
 */
export function createAfdSchemaCommand(
	getCommands: GetDescribedCommands,
	getJsonSchema?: (cmd: DescribedCommand) => Record<string, unknown>
): ZodCommandDefinition<typeof inputSchema, SchemaOutput> {
	return defineCommand({
		name: 'afd-schema',
		description: 'Export input schemas for all commands as JSON Schema or TypeScript types',
		category: 'bootstrap',
		tags: ['bootstrap', 'read', 'safe'],
		mutation: false,
		version: '1.0.0',
		expose: { mcp: true },
		input: inputSchema,

		async handler(input, context) {
			const commands = describableCommands(getCommands, context);

			const schemas: SchemaInfo[] = commands.map((cmd) => {
				let schema: Record<string, unknown>;
				if (getJsonSchema) {
					schema = getJsonSchema(cmd);
				} else if (cmd.jsonSchema) {
					schema = { ...cmd.jsonSchema };
				} else {
					schema = schemaFromParameters(cmd);
				}
				return { name: cmd.name, description: cmd.description, inputSchema: schema };
			});

			if (input.format === 'typescript') {
				const output: SchemaOutput = {
					schemas,
					count: schemas.length,
					format: 'typescript',
					typescript: generateInputTypes(schemas),
				};
				return success(output, {
					reasoning: `Exported ${schemas.length} schemas with TypeScript input types`,
					confidence: 1.0,
				});
			}

			const output: SchemaOutput = { schemas, count: schemas.length, format: 'json' };
			return success(output, {
				reasoning: `Exported JSON schemas for ${schemas.length} commands`,
				confidence: 1.0,
			});
		},
	});
}

File: python/src/afd/server/bootstrap/afd_schema.py
"""``afd-schema`` bootstrap command."""

from __future__ import annotations

from typing import Any, Callable, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, ValidationError

from afd.core.commands import CommandContext, CommandDefinition, CommandParameter
from afd.core.result import CommandResult, success
from afd.core.wire import WireModel
from afd.server.bootstrap.afd_context import BOOTSTRAP_EXPOSE
from afd.server.validation import _input_validation_failure


class AfdSchemaInput(BaseModel):
    """Input for ``afd-schema``."""

    format: Literal["json", "typescript"] = Field(
        default="json",
        description='Output format: "json" or "typescript"',
    )


class SchemaInfo(WireModel):
    """Schema information for one command."""

    name: str
    description: str
    input_schema: Dict[str, Any]
    output_schema: Dict[str, Any] | None = None
    typescript: str | None = None


class AfdSchemaOutput(WireModel):
    """Output for ``afd-schema``."""

    schemas: List[SchemaInfo]
    count: int
    format: Literal["json", "typescript"]


def _build_schema_from_parameters(parameters: List[CommandParameter]) -> Dict[str, Any]:
    properties: Dict[str, Dict[str, Any]] = {}
    required: List[str] = []
    for parameter in parameters:
        prop = dict(parameter.schema or {})
        prop.setdefault("type", parameter.type)
        if parameter.description:
            prop.setdefault("description", parameter.description)
        if parameter.default is not None:
            prop.setdefault("default", parameter.default)
        if parameter.enum is not None:
            prop.setdefault("enum", parameter.enum)
        properties[parameter.name] = prop
        if parameter.required:
            required.append(parameter.name)

    return {
        "type": "object",
        "properties": properties,
        "required": required,
    }


def _schema_to_typescript(schema: Dict[str, Any] | None) -> str:
    if not schema:
        return "unknown"

    if schema.get("const") is not None:
        return repr(schema["const"])

    if schema.get("enum"):
        return " | ".join(repr(value) for value in schema["enum"])

    if schema.get("oneOf"):
        return " | ".join(_schema_to_typescript(item) for item in schema["oneOf"])

    if schema.get("anyOf"):
        return " | ".join(_schema_to_typescript(item) for item in schema["anyOf"])

    if schema.get("allOf"):
        return " & ".join(_schema_to_typescript(item) for item in schema["allOf"])

    schema_type = schema.get("type")
    if isinstance(schema_type, list):
        return " | ".join(_schema_to_typescript({"type": item}) for item in schema_type)
    if schema_type == "string":
        return "string"
    if schema_type == "number" or schema_type == "integer":
        return "number"
    if schema_type == "boolean":
        return "boolean"
    if schema_type == "null":
        return "null"
    if schema_type == "array":
        return f"Array<{_schema_to_typescript(schema.get('items'))}>"
    if schema_type == "object":
        properties = schema.get("properties", {}) or {}
        required = set(schema.get("required", []) or [])
        if not properties:
            return "Record<string, unknown>"
        parts = ["{"]
        for name, prop_schema in properties.items():
            optional = "" if name in required else "?"
            parts.append(f"  {name}{optional}: {_schema_to_typescript(prop_schema)};")
        parts.append("}")
        return "\n".join(parts)
    return "unknown"


def _command_type_name(command_name: str) -> str:
    return "".join(part.capitalize() for part in command_name.split("-"))


def _render_named_typescript_type(name: str, schema: Dict[str, Any] | None) -> str:
    rendered = _schema_to_typescript(schema)
    if schema and schema.get("type") == "object" and rendered.startswith("{"):
        return f"export interface {name} {rendered}"
    return f"export type {name} = {rendered};"


def _render_typescript(command: CommandDefinition, input_schema: Dict[str, Any], output_schema: Dict[str, Any] | None) -> str:
    base_name = _command_type_name(command.name)
    return "\n".join(
        [
            _render_named_typescript_type(f"{base_name}Input", input_schema),
            _render_named_typescript_type(f"{base_name}Output", output_schema),
        ]
    )


async def _afd_schema_handler(
    input: AfdSchemaInput,
    context: Optional[CommandContext],
    get_commands: Callable[[], List[CommandDefinition]],
    get_json_schema: Optional[Callable[[CommandDefinition], Dict[str, Any]]] = None,
) -> CommandResult[AfdSchemaOutput]:
    commands = get_commands()
    schemas: list[SchemaInfo] = []

    for command in commands:
        input_schema = (
            get_json_schema(command)
            if get_json_schema is not None
            else command.input_schema or _build_schema_from_parameters(command.parameters)
        )
        output_schema = command.returns
        typescript = None
        if input.format == "typescript":
            typescript = _render_typescript(command, input_schema, output_schema)

        schemas.append(
            SchemaInfo(
                name=command.name,
                description=command.description,
                input_schema=input_schema,
                output_schema=output_schema,
                typescript=typescript,
            )
        )

    return success(
        AfdSchemaOutput(schemas=schemas, count=len(schemas), format=input.format),
        reasoning=f"Exported {len(schemas)} command schemas in {input.format} format",
        confidence=1.0,
    )


def create_afd_schema_command(
    get_commands: Callable[[], List[CommandDefinition]],
    get_json_schema: Optional[Callable[[CommandDefinition], Dict[str, Any]]] = None,
) -> CommandDefinition:
    """Create the ``afd-schema`` bootstrap command."""

    async def handler(
        input: Any,
        context: Optional[CommandContext] = None,
    ) -> CommandResult[AfdSchemaOutput]:
        try:
            parsed_input = input if isinstance(input, AfdSchemaInput) else AfdSchemaInput(**(input or {}))
        except ValidationError as exc:
            return _input_validation_failure(AfdSchemaInput, input, exc)
        return await _afd_schema_handler(parsed_input, context, get_commands, get_json_schema)

    return CommandDefinition(
        name="afd-schema",
        description="Export input and output schemas for available commands",
        handler=handler,
        category="bootstrap",
        tags=["bootstrap", "read", "safe"],
        mutation=False,
        version="1.0.0",
        parameters=[
            CommandParameter(
                name="format",
                type="string",
                description='Output format: "json" or "typescript"',
                required=False,
                enum=["json", "typescript"],
            )
        ],
        input_schema=AfdSchemaInput.model_json_schema(),
        returns=AfdSchemaOutput.model_json_schema(),
        expose=BOOTSTRAP_EXPOSE,
    )
